cmake/information_helpers.cmake

```diff
@@ -10,67 +12,65 @@ function(filter_generator_expressions INPUT OUTPUT)
     string(REGEX REPLACE "[$<A-Z_:]*SHELL:(.+)>*" "\\1" TMP "${TMP}")
     string(REGEX REPLACE ".+INTERFACE:.+>" "" TMP "${TMP}")
     string(REGEX REPLACE "\$<COMMA>" "," TMP "${TMP}")
+    # Remove the left : or > 
+    string(REGEX REPLACE ":|>" "" TMP "${TMP}")
     # Ignore hwloc include if it is the internal one
     string(REGEX REPLACE "${PROJECT_BINARY_DIR}.*hwloc/src/include.*" "" TMP "${TMP}")
     set(${OUTPUT} "${TMP}" PARENT_SCOPE)
 endfunction()
 
 macro(ginkgo_interface_libraries_recursively INTERFACE_LIBS)
+    # always add the interface to the list to keep the order information 
+    # Currently, it does not support the circular dependence and MSVC.
     foreach(_libs ${INTERFACE_LIBS})
-        if (NOT "${_libs}" IN_LIST GINKGO_INTERFACE_LIBS_FOUND
-                AND NOT "-l${_libs}" IN_LIST GINKGO_INTERFACE_LIBS_FOUND)
-            if (TARGET ${_libs})
-                if (upper_CMAKE_BUILD_TYPE STREQUAL "DEBUG" AND "${_libs}" MATCHES "ginkgo.*")
-                    set(GINKGO_INTERFACE_LIB_NAME "-l${_libs}${CMAKE_DEBUG_POSTFIX}")
-                elseif("${_libs}" MATCHES "ginkgo.*") # Ginkgo libs are appended in the form -l
-                    set(GINKGO_INTERFACE_LIB_NAME "-l${_libs}")
+        if (TARGET ${_libs})
+            if("${_libs}" MATCHES "ginkgo.*")
+                set(GINKGO_INTERFACE_LIB_NAME "-l${_libs}")
+                if(upper_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
+                    set(GINKGO_INTERFACE_LIB_NAME "${GINKGO_INTERFACE_LIB_NAME}${CMAKE_DEBUG_POSTFIX}")
                 endif()
-
-                # Get the link flags and treat them
-                get_target_property(GINKGO_INTERFACE_LIBS_LINK_FLAGS "${_libs}"
-                    INTERFACE_LINK_OPTIONS)
-                if (GINKGO_INTERFACE_LIBS_LINK_FLAGS)
-                    filter_generator_expressions("${GINKGO_INTERFACE_LIBS_LINK_FLAGS}"
-                        GINKGO_INTERFACE_LIB_NAME)
-                endif()
-                if (NOT "${GINKGO_INTERFACE_LIB_NAME}" IN_LIST GINKGO_INTERFACE_LIBS_FOUND)
-                    list(APPEND GINKGO_INTERFACE_LIBS_FOUND "${GINKGO_INTERFACE_LIB_NAME}")
+                list(APPEND GINKGO_INTERFACE_LIBS_FOUND "${GINKGO_INTERFACE_LIB_NAME}")
```

Comment:
can we do this with a generator expression? `$<$<CONFIG:Debug>:${CMAKE_DEBUG_POSTFIX}>`
This would also make things work in multi-config generators